lib/prettyprint.js

```diff
@@ -287,11 +287,61 @@ function prettyprintProgram(ast, short) {
             ast.rules.map(prettyprintRule).join('') + '}';
 }
 
+function prettyprintExample(ex) {
+    let args = Object.keys(ex.args);
+    let types = args.map((a) => ex.args[a]);
+
+    let arg_decl = args.length > 0 ? `(${args.map((a, i) => a + ' :' + prettyprintType(types[i])).join(', ')})` : '';
+
+    let value;
+    switch (ex.type) {
+        case 'stream':
+            value = prettyprintStream(ex.value);
+            break;
+        case 'query':
+            value = prettyprintTable(ex.value);
+            break;
+        case 'action':
+            value = prettyprintAction(ex.value);
+            break;
+        case 'program':
+            value = prettyprintProgram(ex.value);
+            break;
+        default:
+            throw new TypeError();
+    }
+    let annotation = `#_[utterances=[${ex.utterances.map((u) => prettyprintValue(u)).join(',')}]]`;
+
+    return `${ex.type} ${arg_decl} := ${value} ${annotation};\n`;
+}
+
+function prettyprintDataset(ast) {
+    let examples = '';
+    let prefix = '    ';
+    ast.examples.forEach((ex) => {
+        examples += prefix + prettyprintExample(ex);
+    });
+    return `dataset ${ast.name} language '${ast.language}' {\n${examples}}`;
+}
+
+function prettyprintMeta(ast) {
+    let meta = '';
+    for (let klass of ast.classes)
+        meta += prettyprintClassDef(klass);
+    for (let dataset of ast.datasets)
+        meta += prettyprintDataset(dataset);
+    return meta;
+}
+
 function prettyprint(ast, short) {
     if (ast.isProgram)
         return prettyprintProgram(ast, short);
-    else
+    else if (ast.isPermissionRule)
         return prettyprintPermissionRule(ast, short);
+    else if (ast.isMeta)
+        return prettyprintMeta(ast, short);
+    else
+        return prettyprintDataset(ast, short);
```

Comment:
Wrong merge here? A dataset is not an input on its own, right?

Reply:
fixed